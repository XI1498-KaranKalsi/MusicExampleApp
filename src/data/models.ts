export interface Song {
  wrapperType?: 'audiobook' | 'track',
  artistId?: number;
  collectionId?: number;
  artistName: string;
  collectionName?: string;
  collectionCensoredName?: string;
  artistViewUrl?: string;
  collectionViewUrl?: string;
  artworkUrl60: string;
  artworkUrl100: string;
  collectionPrice: number;
  trackCount?: number;
  copyright?: string;
  releaseDate: string;
  primaryGenreName: string;
  previewUrl: string;
  description?: string;
  amgArtistId?: number;
  trackId?: number;
  trackName?: string;
  trackCensoredName?: string;
  trackViewUrl?: string;
  artworkUrl30?: string;
  trackPrice?: number;
  discCount?: number;
  discNumber?: number;
  trackNumber?: number;
  trackTimeMillis?: number;
  contentAdvisoryRating?: string;
  isStreamable?: boolean;
  collectionArtistId?: number;
  collectionArtistName?: string;
  collectionArtistViewUrl?: string;
  trackRentalPrice?: number;
  collectionHdPrice?: number;
  trackHdPrice?: number;
  trackHdRentalPrice?: number;
  longDescription?: string;
  hasITunesExtras?: boolean;
  shortDescription?: string;
}

export interface SearchSongsResponse {
  resultCount: number;
  results: Song[];
}
