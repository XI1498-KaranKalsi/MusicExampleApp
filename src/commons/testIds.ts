const testIds = {
    SongsList: {
        Loader: 'SongList_Loader',
        Flatlist: 'SongList_FlatList'
    },
    SongItem: {
        Title: 'SongItem_Title',
        Artist: 'SongItem_Artist',
        Artwork: 'SongItem_Artwork'
    }
}

export default testIds;