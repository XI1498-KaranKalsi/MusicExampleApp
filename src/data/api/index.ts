import wretch, {ResponseChain, Wretcher} from 'wretch';
import {SearchSongsResponse, Song} from '../models';
import endpoints from './endpoints';

export const BASE_URL = 'https://itunes.apple.com/';

export const apiUrl = (endpoint: string) => `${BASE_URL}${endpoint}`;

const searchSongs = async (term: string): Promise<Song[]> => {
  const res: ResponseChain = await wretch(apiUrl(endpoints.searchSongs))
    .query({
      term,
    })
    .get();
  const data = await res.json<SearchSongsResponse>();
  return data.results.filter(song=>song.wrapperType === 'track');
};

export default {
  searchSongs,
};
