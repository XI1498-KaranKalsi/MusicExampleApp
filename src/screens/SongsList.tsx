import React from 'react';
import {
  ActivityIndicator,
  ActivityIndicatorProps,
  FlatList,
  Text,
  View,
} from 'react-native';
import {
  NavigationComponentProps,
  NavigationFunctionComponent,
} from 'react-native-navigation';
import colors from '../commons/colors';
import fonts from '../commons/fonts';
import styles from '../commons/styles';
import testIds from '../commons/testIds';
import SongItem from '../components/SongItem';
import api from '../data/api';
import {Song} from '../data/models';
import AppNavigation from '../lib/navigation';
import SongListStyles from './styles/SongListStyles';

export const SONGS_SEARCH_TERM = 'Michael jackson';

export interface SongsListProps extends NavigationComponentProps {}

const Loader = (props: ActivityIndicatorProps): React.ReactElement => {
  return props.animating ? (
    <View style={SongListStyles.loader}>
      <ActivityIndicator {...props} animating={props.animating} />
      <Text style={SongListStyles.loaderText}>Please wait</Text>
    </View>
  ) : (
    <></>
  );
};

type LoadSongsEffectResponse = [boolean, Song[] | undefined];

const loadSongsEffect = (deps: any[]): LoadSongsEffectResponse => {
  const [isSongsLoading, setIsSongsLoading] = React.useState<boolean>(false);
  const [songs, setSongs] = React.useState<Song[]>([]);
  React.useEffect(() => {
    let mounted = true;
    const fetchData = async () => {
      setIsSongsLoading(true);
      const songs = await api.searchSongs(SONGS_SEARCH_TERM);
      if (mounted) {
        setSongs(songs);
        setIsSongsLoading(false);
      }
    };
    fetchData();
    return () => {
      mounted = false;
    };
  }, deps);

  return [isSongsLoading, songs];
};

const navigateToDetail = (componentId: string, song: Song) => {
  AppNavigation.push(componentId, {
    component: {
      name: 'SongDetail',
      passProps: {
        song,
      },
    },
  });
};

const SongsList: NavigationFunctionComponent<SongsListProps> = props => {
  const [isLoading, songs] = loadSongsEffect([]);

  const onSongPressCallback = React.useCallback(
    (song: Song) => navigateToDetail(props.componentId, song),
    [props.componentId],
  );

  return (
    <View style={styles.defaultScreenContainer}>
      <FlatList<Song>
        testID={testIds.SongsList.Flatlist}
        style={SongListStyles.songsList}
        data={songs}
        renderItem={({item}) => (
          <SongItem song={item} onSongPress={onSongPressCallback} />
        )}
        keyExtractor={item =>
          `${item.trackId}${item.collectionId}${item.artistId}`
        }
      />
      <Loader
        testID={testIds.SongsList.Loader}
        color={colors.primaryDark}
        animating={isLoading}
      />
    </View>
  );
};

SongsList.options = {
  topBar: {
    title: {
      color: colors.accent,
      alignment: 'center',
      text: 'Songs',
    },
  },
};

export default SongsList;
