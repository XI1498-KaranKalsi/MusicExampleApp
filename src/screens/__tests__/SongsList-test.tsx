import React from 'react';
import renderer from 'react-test-renderer';
import {cleanup, render, waitFor} from '@testing-library/react-native';
import fetchMock from 'fetch-mock';
import SongsList, {SONGS_SEARCH_TERM, SongsListProps} from '../SongsList';
import {apiUrl} from '../../data/api';
import endpoints from '../../data/api/endpoints';
import searchSongsStub from '../../stub/searchSongs.json';
import testIds from '../../commons/testIds';

describe('testing screen SongsList', () => {
  const defaultProps: SongsListProps = {
    componentId: 'Component1',
  };

  beforeEach(() => {
    fetchMock.reset();
  });

  afterEach(() => {
    cleanup();
  });

  it('rendering correctly', () => {
    expect(renderer.create(<SongsList {...defaultProps} />)).toBeTruthy();
  });
  it('matches with snapshot', () => {
    const {toJSON} = render(<SongsList {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });
  it('data loading correctly', async () => {
    fetchMock.get(
      apiUrl(endpoints.searchSongs),
      {
        status: 200,
        body: searchSongsStub,
      },
      {
        query: {
          term: SONGS_SEARCH_TERM,
        },
        delay: 100,
      },
    );

    const {getByTestId, queryByTestId} = render(
      <SongsList {...defaultProps} />,
    );

    expect(
      getByTestId(testIds.SongsList.Loader)?.props['animating'],
    ).toBeTruthy();
    expect(getByTestId(testIds.SongsList.Flatlist)?.props['data'].length).toBe(
      0,
    );
    await waitFor(() =>
      expect(
        queryByTestId(testIds.SongsList.Loader)?.props['animating'],
      ).toBeFalsy(),
    );
    expect(
      getByTestId(testIds.SongsList.Flatlist)?.props['data'].length,
    ).toBeGreaterThan(0);
  });
});
