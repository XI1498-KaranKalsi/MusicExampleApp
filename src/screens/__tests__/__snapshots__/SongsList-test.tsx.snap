// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing screen SongsList matches with snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e3f0fc",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "width": "100%",
      }
    }
    testID="SongList_FlatList"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "position": "absolute",
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#2596be"
      hidesWhenStopped={true}
      size="small"
      testID="SongList_Loader"
    />
    <Text
      style={
        Object {
          "color": "#2596be",
          "fontSize": 14,
          "marginTop": 10,
        }
      }
    >
      Please wait
    </Text>
  </View>
</View>
`;
